### PR TITLE
Various `cabal-testsuite` improvements

### PR DESCRIPTION
**Pull request by jasagredo**
- Make `.bat` scripts for CCompilerOverride test look into the shipped toolchain for gcc/clang.
- The `flaky` combinator can be used to specify flaky tests with a ticket number. These will be reported as passing or failing but will not make the test-suite error.
- Broken tests can be broken by accept-test, i.e. by the output not matching. Previously a test that failed this way had to be skipped because it didn't pass but the expectBroken logic would deem it not broken.
- Also makes `postCheckoutCommand` not flaky on Windows.
- It removes some outdated tests (old GHCs) and some tests that will never pass again.
- Fixes some issues with output normalization in MacOS

Depends-On: #10282

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

File: cabal-testsuite/src/Test/Cabal/Monad.hs
Comment (jasagredo) [inline cabal-testsuite/src/Test/Cabal/Monad.hs:63]: Should move these somewhere else

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (jasagredo) [inline cabal-testsuite/src/Test/Cabal/TestCode.hs:50]: Should I use more concrete datatypes than `Bool`?

Comment (geekosaur) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs:50]: Honestly, the tuple bothers me more.

Comment (ulysses4ever) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs:50]: Yeah, please, let's have something more descriptive than `(Bool, Int)`. While on it, how about introducing an alias `type IssueID = Int`?

E.g.

```haskell
type TestPassed = Bool
type IssueID = Int
data FlakyStatus
  = NotFlaky
  | Flaky TestPassed IssueID

isTestCodeFlaky :: TestCode -> FlakyStatus
```
(haven't checked if `fourmolu` would approve this indentation)

I see they used `Maybe` below for a similar use case, but I'd say it's a code smell.

Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs:50]: I'm adding some more stuff I thought about this evening. Will push something later, I think it will look much better. Thanks both

Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs]: Pushed. The commits should show a comprehensive story about the changes.

**Comment by jasagredo** [comment]
Hm what a surprising failure on macOS! Will investigate

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/TestCode.hs:26]: Could you use the `IssueID` alias here and in other places where it makes sense?

File: cabal-testsuite/src/Test/Cabal/Monad.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/Monad.hs:59]: You will need to update the documentation in https://github.com/haskell/cabal/tree/master/cabal-testsuite#frequently-asked-questions

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/Prelude.hs:999]: I am not sure I buy it as an improvement having separate functions for so many combinations. It'd be great to have a little DSL for expressing such things and this looks like a step away from that. But it is tempting to go this route because it does save on typing. Perhaps if Haskell had a built-in the monadic bang (in the sense of [this plugin](https://hackage.haskell.org/package/monadic-bang)), the savings here wouldn't be as noticeable. So, overall, I don't feel strongly about it, just curious what others think.

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs:26]: Ah yes indeed, will do

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:999]: Yeah, as I was writing them I was thinking the same. But overall I think it is useful to make the skip and broken messages as uniform as possible, to be able to grep the output. 

Otherwise one might use "shared libs", another person uses "dynamic libs", another one makes a typo "shred libs", etc.

I will consider whether to add something like a DSL. But maybe this would be mergeable? At the very least I could put a comment in the code saying "TODO: this should be reworked into a DSL or some such".

Comment (geekosaur) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:999]: You don't need to risk loss of uniformity. See for example https://hackage.haskell.org/package/xmonad-contrib-0.18.0/docs/XMonad-Util-WindowProperties.html.
In this case I'd have 4 parameters: expected state or action (skip, expect fail, expect pass, flaky…), predicate (cf. above WindowProperties, replacing your multiple conditionals), ticket, message. But this might indeed be future work.

Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:999]: Ah i didn't explain myself. I think a DSL would be uniform and the best solution. What is not uniform is
```
skipIf noShared "no shared libs"
```
And somewhere else
```
skipIf noShared "no dyn libs"
```

Which is what was there before I introduced these functions.

Comment (geekosaur) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:999]: I would make that aspect of the message come from the DSL, and the explanation should give details. (In which case perhaps it should be optional, or absorbed into the ticket.)

**Comment by jasagredo** [comment]
I think I fixed the MacOS issue with 4285454efcb9d0396e31b5ba324af5c55a37c0fd. Let's see what CI says, then I will address the review comments.

For the DSL, I tried doing something last night but things started looking very bad as there were just too many edge cases and conditions applied in different places. I will give it some more thought.

File: cabal-testsuite/src/Test/Cabal/Monad.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Monad.hs:59]: Done

File: cabal-testsuite/README.md
Comment (jasagredo) [inline cabal-testsuite/README.md:212]: For the reviewer, now it **does** handle that.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:999]: Another option I could see is by making the combinators return a `Maybe String` and then `skipIf` would unpack it if `Just`, something like

```
isWindows :: Maybe String
isWindows = if buildOS == Windows then Just "Windows" else Nothing

skipIf mSkipMsg test = maybe test skip mSkipMsg
```

but that is counterintuitive. The suffix `if` usually is interpreted as expecting a `Bool`...

**Comment by jasagredo** [comment]
The `RejectFutureIndexStates` test is timing-out on Windows on CI. This doesn't happen on my local machine, which is kind of weird.

I will probably have to mark it as Skip on Windows+CI 🙄 or investigate with the tmate action.

**Comment by jasagredo** [comment]
I deferred the investigation of those two tests to #10230. Before this PR they were skipped anyway on all OSes.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:1002]: Ok I managed to make the CI green. From my side there are no other blockers for merging this.

What is your opinion on the combinators then @ulysses4ever @geekosaur ? Could this be merged and the DSL be done in a later PR or do you think this should not be merged before the DSL is done?

Comment (ulysses4ever) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:1002]: By no means I meant that the DSL should be a prerequisite.

I simply struggle to get to the bottom of the long line of commits. For me personally, it'd be much easier to approve commits in separate PRs one by one. But I understand how inconvenient that approach is for the author. So I'll try to get a review done.

File: Cabal-tests/lib/Test/Utils/TempTestDir.hs
Comment (ulysses4ever) [inline Cabal-tests/lib/Test/Utils/TempTestDir.hs:42]: The comment doesn't describe the added `when` does it? It just explains why `removeDIrectoryRecursiveHack` is there? But then I'm not sure the comment adds anything to the top-level comment for `withTempDir`... Or am I missing something?

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:1002]: Some points I wanted to note down:
- It would have been perfectly ok if you said you wanted the DSL before approving, it is a perfectly valid opinion and I asked because I would have thought it was perfectly ok if you said "I will not give my approval until this PR has better ergonomics for the combinators like a DSL maybe"
- I have tried to split the changes semantically, one on each commit, so that it would be easier to review. Some of them are interesting, some are boring 😄
- (Btw, while reviewing you can choose to see only one commit at a time, that might make it easier)
- I'm fine with splitting the work in multiple PRs if it is independent (kind of like in this case, it could have been 4-5 PRs) but I think it is more work to drag attention to 5 PRs for review than to one. There is usually not enough volunteer-time to review this so I prefer just making it in one go 🙂 (as long as the commits are related, like in this case they are all about CI and tests)
- This being said, if you prefer I can split this PR in multiple PRs.

File: Cabal-tests/lib/Test/Utils/TempTestDir.hs
Comment (jasagredo) [reply Cabal-tests/lib/Test/Utils/TempTestDir.hs:42]: The `when` comes from the change I made for tests to use this instead of `withTempDirectoryEx`. It is the key for the tests to not delete their temporal directory.

Probably I should write that down in a comment. Also I missed the other comment you mention, what I wrote seems indeed redundant.

File: cabal-testsuite/src/Test/Cabal/Monad.hs
Comment (jasagredo) [inline cabal-testsuite/src/Test/Cabal/Monad.hs:885]: For the reviewer, this change is crucial, because it allows us to manipulate the output in the running directory, like we do in RejectFutureIndexStates, to replace the REPLACEME string in the output with the returned timestamp.

Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Monad.hs:885]: Probably I should note that down in a comment.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (geekosaur) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:1002]: I think I already suggested it should be future work? It wants some up-front design.

File: Cabal-tests/lib/Test/Utils/TempTestDir.hs
Comment (ulysses4ever) [reply Cabal-tests/lib/Test/Utils/TempTestDir.hs:42]: Good, that's what I thought. Yeah, adding a comment about temp dirs and (re)moving duplication would be great.

File: cabal-testsuite/src/Test/Cabal/Prelude.hs
Comment (ulysses4ever) [reply cabal-testsuite/src/Test/Cabal/Prelude.hs:1002]: For splitting: I think you should do it the way you prefer. For me personally reviewing and approving 5 small PRs can take much less time than reviewing 5 commits and approving one PR  with those. That's simply because it's not possible to "save state" (meaning my mental state) in between the commits while there's a way to do that for multiple PRs (by approving them one by one). Unless I'm missing something in GitHub UI. I know that you can review commits separately (that's what I've been doing with this PR), but that's not the same as reviewing/approving/and forgetting about individual PRs (which works better for me). Again, you should do it the way you find it more convenient, I think.

For the DSL, I share Brandon's opinion (see the comment just above) entirely.

File: Cabal-tests/lib/Test/Utils/TempTestDir.hs
Comment (jasagredo) [reply Cabal-tests/lib/Test/Utils/TempTestDir.hs]: Done

**Comment by jasagredo** [comment]
May I ask for an approval if there are no further comments @ulysses4ever ? (Just pinging you because you had reviewed the pr already, I can ask in matrix if you can't review it)

**Comment by ulysses4ever** [comment]
@jasagredo it's on my radar, yes. In the meantime, applying the needs-review label (just did it) and asking on Matrix may speed it up. I can't guarantee that I get to it today but I'll try my best.

**Comment by ulysses4ever** [comment]
@mpickering I think you were one of the last people to do a meaningful update to the Cabal testsuite. Could you, perhaps, take a look at this PR?

File: cabal-testsuite/main/cabal-tests.hs
Comment (ulysses4ever) [inline cabal-testsuite/main/cabal-tests.hs:255]: This grepping through stdout feels a little crude although I don't know a better way. Also, usually, when you switch from showing direct output to showing captured one, there are some side effects. E.g. colored output would be unavailable, etc. Of course, Cabal testsuite doesn't have colored output in particular but there may be spooky actions at the distance...

File: cabal-testsuite/src/Test/Cabal/Monad.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/Monad.hs:232]: Can we change it to `printExpectedBroken` or `expectedBrokenMsg` or anything like this? The current name is just too close to `expectBroken` while serving a completely different purpose. I know you're not the author of these names but since you're updating it anyway...

Same for the other printing functions.

Also since you're updating this: could you replace `Int` with the type synonym for issue IDs?

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/TestCode.hs:60]: Highjacking `Show` for pretty printing is considered a code smell. Why not a regular function?

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/TestCode.hs:62]: Apart from adhering to the roundtripping law for `read.show`, is there a reason to define this? If not, as I said above, you could probably avoid `Show` and, therefore, avoid this instance.

File: cabal-testsuite/main/cabal-tests.hs
Comment (jasagredo) [reply cabal-testsuite/main/cabal-tests.hs:255]: I thought maybe there was some other way to pipe this output of the process but I didn't seem to find one.

In any case this is only for when you invoke the tests with one test. I don't know how much that is used, I personally run the validate script always

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs:62]: The test code output is read by the testing ghci servers to determine whether a test passes or not. Unless I defined this it would say that it can't parse some outcomes.

But I think show could be the stock show and I could pretty show with some other function.

File: cabal-testsuite/main/cabal-tests.hs
Comment (ulysses4ever) [reply cabal-testsuite/main/cabal-tests.hs:255]: > this is only for when you invoke the tests with one test

Oh, that's much less of a concern then. Thank you!

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (ulysses4ever) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs:62]: Right… Yeah, of there was a way to avoid touching Show/Read, I'd prefer that. It's probably less code too. But if not, that's not the end of the world of course.

File: cabal-testsuite/main/cabal-tests.hs
Comment (ulysses4ever) [reply cabal-testsuite/main/cabal-tests.hs:255]: Let's leave it as unresolved for now maybe to see if anyone has other thoughts...

File: cabal-testsuite/src/Test/Cabal/Monad.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/Monad.hs]: Done

File: cabal-testsuite/src/Test/Cabal/TestCode.hs
Comment (jasagredo) [reply cabal-testsuite/src/Test/Cabal/TestCode.hs]: Removed

**Comment by ulysses4ever** [comment]
[Remove outdated tests](https://github.com/haskell/cabal/pull/10225/commits/aa38d64173143413a823759d5f2885357cadb129) -- for each test it'd be good to have a short phrase why it's considered outdated. I know this is a little daunting of a task, and I won't be blocking the PR on it.

File: cabal-testsuite/PackageTests/NewUpdate/UpdateIndexState/update-index-state.test.hs
Comment (ulysses4ever) [inline cabal-testsuite/PackageTests/NewUpdate/UpdateIndexState/update-index-state.test.hs]: Ibid.

File: cabal-testsuite/PackageTests/NewSdist/DeterministicTrivial/deterministic.test.hs
Comment (ulysses4ever) [inline cabal-testsuite/PackageTests/NewSdist/DeterministicTrivial/deterministic.test.hs:27]: I approve this with a heavy heart because of how much pain "DeterministicTrivial" has brought us due to its flakiness in the past (what a name for such a record!..). I truly hope that it's fixed otherwise we're doooomed...

File: cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs:141]: I don't understand this comment.

**Review by ulysses4ever — APPROVED** [review]
Thank you and sorry it took so long!

**Comment by geekosaur** [comment]
Are we ready to set the merge label on this? (Squash?)

**Comment by jasagredo** [comment]
I need to address @ulysses4ever 's remarks and this should be ready to merge.

**Comment by jasagredo** [comment]
Let's wait until the Windows CI is reenabled before merging this

**Comment by geekosaur** [comment]
I set a dependency so Mergify should merge it automatically when #10282 goes in.

**Pull request merged**
